expo/expo-cli
# [xdl] Add robots as new supported user type

## What I did

This integrates the Robot users, as new user type, in the XDL. This PR was built on top of [the `EXPO_TOKEN` integration](https://github.com/expo/expo-cli/pull/2415).

- **Added helper method to find the "owner" of a project**
With the addition of robot users, we can't simply use `manifest.owner || user.username`. Robots don't have usernames and _must_ be used with the manifest owner property. I added a `getProjectOwner` method to "safely" fetch the project owner, and fail if there is none set when using robot users. I think I replaced all "unsafe" owner fetches, but this is good to double-check.

- **Added XDL User methods to only fetch users or robots**
When authenticated as a different user type, the CLI will exit with "This action only supports user type X". Useful to quickly block robot users on unsupported endpoints

- **Test all commands**
Done, see [this doc](https://www.notion.so/expo/Bot-Access-CLI-5c4d04e9034d45b9b9faeea106eaf7e4#6de4705942af4c4ab4f9632085855370).

## What still needs to be done

- ~~**Current credentials API doesn't support robots**~~
~~This is a blocker for running the `$ expo build:*` commands. We need the credentials to start the builds.~~

- ~~**Move integration to EAS**~~
~~After we finished the integration in Expo CLI, I'll work on moving it over to EAS.~~ 
[See PR #108](https://github.com/expo/eas-cli/pull/108)

# Test Plan

**To test manually**
- Create a robot user for your account (see GraphQL query below, not yet available on website)
- Create an access token for your robot user ([using this link](https://staging.expo.io/settings/access-tokens))
- Try smoke testing commands with `EXPO_TOKEN=<robot-token> expo ...` (e.g. `expo whoami`, [this doc contains all tested commands by me](https://www.notion.so/expo/Bot-Access-CLI-5c4d04e9034d45b9b9faeea106eaf7e4#6de4705942af4c4ab4f9632085855370))

<details>
<summary>Create a robot with graphql mutation</summary>
<pre>
mutation {
  robot {
    createRobotForAccount(
      accountID: "{id}"
      robotData: { name: "GLaDOS" }
      permissions: PUBLISH
    ) {
      id
      firstName
    }
  }
}
</pre>
</details>
